readme.txt
Angband 4.0.5
=============

Angband is a graphical dungeon adventure game that uses textual characters
to represent the walls and floors of a dungeon and the inhabitants therein, 
in the vein of games like NetHack and Rogue.  If you need help in-game,
press '?'.

For more information, somewhere to upload your characters and screenshots,
and discuss the game, try http://angband.oook.cz/.

If you're compiling the game yourself, read http://trac.rephial.org/wiki/Compiling.

Enjoy!

-- The Angband Dev Team

---------------------------------------

To-do (so far, this branch (a reminder for myself)):

Fix Reduce Cuts and Poison, add a message for TIMED_DEC:CUT so it is clear that it has reduced cuts.

Fix Resist Magic, TIMED_INVULN has an entirely different effect in Vanilla.

Add Magic Book for Mages #2.

Add all the other books.


lib/history.txt
# File: history.txt


# This file is used to initialize the "player history" information for 
# the Angband game.

# Do not modify this file unless you know exactly what you are doing,
# unless you wish to risk possible system crashes and broken savefiles.

# Background information:
# Pieces of player backstory are chosen from "charts", which are lists of
# phrases or sentences with similar general meaning.

# 'chart' lines give the number of the current chart, the number of the chart
# that will be used next (or 0 if the last chart), and a cutoff for choosing
# the current line.  If the first field in two chart lines is the same, the
# second should be as well, to maintain consistent choices and end up with
# something readable.

# 'phrase' lines are the actual chart entries which make up the history.

# Chart progression by race:
#   Human/Dunadan -->  1 -->  2 -->  3 --> 50 --> 51 --> 52 --> 53
#   Half-Elf      -->  4 -->  1 -->  2 -->  3 --> 50 --> 51 --> 52 --> 53
#   Elf           -->  5 -->  6 -->  9 --> 54 --> 55 --> 56
#   High-Elf      -->  7 -->  8 -->  9 --> 54 --> 55 --> 56
#   Hobbit        --> 10 --> 11 -->  3 --> 50 --> 51 --> 52 --> 53
#   Gnome         --> 13 --> 14 -->  3 --> 50 --> 51 --> 52 --> 53
#   Dwarf         --> 16 --> 17 --> 18 --> 57 --> 58 --> 59 --> 60 --> 61
#   Half-Orc      --> 19 --> 20 -->  2 -->  3 --> 50 --> 51 --> 52 --> 53
#   Half-Troll    --> 21 --> 22 --> 62 --> 63 --> 64 --> 65 --> 66
#   Kobold        --> 23 --> 24 --> 25 --> 26 --> 67 --> 68 --> 69

# This table *must* be correct or the history may look very silly.

# Note that the "spacing" in the "description" lines is very important!



chart:1:2:10
phrase:You are the illegitimate and unacknowledged child 
chart:1:2:20
phrase:You are the illegitimate but acknowledged child 
chart:1:2:95
phrase:You are one of several children 
chart:1:2:100
phrase:You are the first child 

chart:2:3:40
phrase:of a Serf.  
chart:2:3:65
phrase:of a Yeoman.  
chart:2:3:80
phrase:of a Townsman.  
chart:2:3:90
phrase:of a Guildsman.  
chart:2:3:96
phrase:of a Landed Knight.  
chart:2:3:99
phrase:of a Titled Noble.  
chart:2:3:100
phrase:of a Royal Blood Line.  

chart:3:50:20
phrase:You are the black sheep of the family.  
chart:3:50:80
phrase:You are a credit to the family.  
chart:3:50:100
phrase:You are a well liked child.  

chart:4:1:15
phrase:Your mother was of the Avari. 
chart:4:1:30
phrase:Your father was of the Avari. 
chart:4:1:50
phrase:Your mother was of the Nandor. 
chart:4:1:70
phrase:Your father was of the Nandor.  
chart:4:1:85
phrase:Your mother was of the Sindar.  
chart:4:1:95
phrase:Your father was of the Sindar.  
chart:4:1:98
phrase:Your mother was of the Noldor.  
chart:4:1:100
phrase:Your father was of the Noldor.  

chart:5:6:60
phrase:You are one of several children 
chart:5:6:100
phrase:You are the only child 

chart:6:9:40
phrase:of an Avarin 
chart:6:9:70
phrase:of a Nandorin 
chart:6:9:100
phrase:of a Sindarin 

chart:7:8:60
phrase:You are one of several children 
chart:7:8:100
phrase:You are the only child 

chart:8:9:75
phrase:of a Telerin 
chart:8:9:95
phrase:of a Noldorin 
chart:8:9:100
phrase:of a Vanyarin 

chart:9:54:40
phrase:Ranger.  
chart:9:54:70
phrase:Archer.  
chart:9:54:87
phrase:Warrior.  
chart:9:54:95
phrase:Mage.  
chart:9:54:99
phrase:Prince.  
chart:9:54:100
phrase:King.  

chart:10:11:85
phrase:You are one of several children of a Hobbit 
chart:10:11:100
phrase:You are the only child of a Hobbit 

chart:11:3:10
phrase:Bum.
chart:11:3:20
phrase:Burglar.  
chart:11:3:30
phrase:Miller.  
chart:11:3:40
phrase:Tavern Owner.  
chart:11:3:50
phrase:Home Owner.  
chart:11:3:80
phrase:Warrior.  
chart:11:3:95
phrase:Shirriff.  
chart:11:3:99
phrase:Mayor.  
chart:11:3:100
phrase:Clan Elder.  

chart:13:14:85
phrase:You are one of several children of a Gnome 
chart:13:14:100
phrase:You are the only child of a Gnome 

chart:14:3:20
phrase:Beggar.  
chart:14:3:50
phrase:Braggart.  
chart:14:3:75
phrase:Prankster.  
chart:14:3:95
phrase:Warrior.  
chart:14:3:100
phrase:Mage.  

chart:16:17:25
phrase:You are one of two children of a Dwarven 
chart:16:17:100
phrase:You are the only child of a Dwarven 

chart:17:18:10
phrase:Thief.  
chart:17:18:25
phrase:Prison Guard.  
chart:17:18:75
phrase:Miner.  
chart:17:18:90
phrase:Warrior.  
chart:17:18:99
phrase:Priest.  
chart:17:18:100
phrase:King.  

chart:18:57:15
phrase:You are the black sheep of the family.  
chart:18:57:85
phrase:You are a credit to the family.  
chart:18:57:100
phrase:You are a well liked child.  

chart:19:20:25
phrase:Your mother was an Orc, 
chart:19:20:100
phrase:Your father was an Orc, 

chart:20:2:20
phrase:and it is acknowledged. You are the adopted child 
chart:20:2:100
phrase:but it is unacknowledged. You are the adopted child 
chart:20:40:100
phrase:breed with Mannish blood by Saruman.  

chart:21:22:30
phrase:Your mother was a Stone-Troll 
chart:21:22:60
phrase:Your father was a Stone-Troll 
chart:21:22:75
phrase:Your mother was a Cave-Troll 
chart:21:22:90
phrase:Your father was a Cave-Troll 
chart:21:22:95
phrase:Your mother was a Water-Troll 
chart:21:22:100
phrase:Your father was a Water-Troll 

chart:22:62:5
phrase:Cook.  
chart:22:62:95
phrase:Warrior.  
chart:22:62:99
phrase:Shaman.  
chart:22:62:100
phrase:Clan Chief.  

chart:23:24:20
phrase:You are the runt of 
chart:23:24:80
phrase:You come from 
chart:23:24:100
phrase:You are the largest of 

chart:24:25:15
phrase:a litter of 3 pups. 
chart:24:25:40
phrase:a litter of 4 pups. 
chart:24:25:70
phrase:a litter of 5 pups. 
chart:24:25:85
phrase:a litter of 6 pups. 
chart:24:25:95
phrase:a litter of 7 pups. 
chart:24:25:100
phrase:a litter of 8 pups. 

chart:25:26:25
phrase:Your father was a fungus farmer, 
chart:25:26:50
phrase:Your father was a hunter, 
chart:25:26:75
phrase:Your father was a warrior, 
chart:25:26:95
phrase:Your father was a shaman, 
chart:25:26:100
phrase:Your father was the tribal chief, 

chart:26:67:20

src/player.h
/**
 * \file player.h
 * \brief Player implementation
 *
 * Copyright (c) 1997 Ben Harrison, James E. Wilson, Robert A. Koeneke
 * Copyright (c) 2011 elly+angband@leptoquark.net. See COPYING.
 * Copyright (c) 2015 Nick McConnell
 *
 * This work is free software; you can redistribute it and/or modify it
 * under the terms of either:
 *
 * a) the GNU General Public License as published by the Free Software
 *    Foundation, version 2, or
 *
 * b) the "Angband licence":
 *    This software may be copied and distributed for educational, research,
 *    and not for profit purposes provided that this copyright and statement
 *    are included in all such copies.  Other copyrights may also apply.
 */

#ifndef PLAYER_H
#define PLAYER_H

#include "guid.h"
#include "obj-properties.h"
#include "object.h"
#include "option.h"


/**
 * Indexes of the various "stats" (hard-coded by savefiles, etc).
 */
enum
{
	#define STAT(a, b, c, d, e, f, g, h) STAT_##a,
	#include "list-stats.h"
	#undef STAT

	STAT_MAX
};

/** Sexes **/

/* Maximum number of player "sex" types (see "table.c", etc) */
#define MAX_SEXES            2

/* Player sex constants (hard-coded by save-files, arrays, etc) */
#define SEX_FEMALE		0
#define SEX_MALE		1

/**
 * Player race and class flags
 */
enum
{
	#define PF(a,b,c) PF_##a,
	#include "list-player-flags.h"
	#undef PF
	PF_MAX
};

#define PF_SIZE                FLAG_SIZE(PF_MAX)

#define pf_has(f, flag)        flag_has_dbg(f, PF_SIZE, flag, #f, #flag)
#define pf_next(f, flag)       flag_next(f, PF_SIZE, flag)
#define pf_is_empty(f)         flag_is_empty(f, PF_SIZE)
#define pf_is_full(f)          flag_is_full(f, PF_SIZE)
#define pf_is_inter(f1, f2)    flag_is_inter(f1, f2, PF_SIZE)
#define pf_is_subset(f1, f2)   flag_is_subset(f1, f2, PF_SIZE)
#define pf_is_equal(f1, f2)    flag_is_equal(f1, f2, PF_SIZE)
#define pf_on(f, flag)         flag_on_dbg(f, PF_SIZE, flag, #f, #flag)
#define pf_off(f, flag)        flag_off(f, PF_SIZE, flag)
#define pf_wipe(f)             flag_wipe(f, PF_SIZE)
#define pf_setall(f)           flag_setall(f, PF_SIZE)
#define pf_negate(f)           flag_negate(f, PF_SIZE)
#define pf_copy(f1, f2)        flag_copy(f1, f2, PF_SIZE)
#define pf_union(f1, f2)       flag_union(f1, f2, PF_SIZE)
#define pf_comp_union(f1, f2)  flag_comp_union(f1, f2, PF_SIZE)
#define pf_inter(f1, f2)       flag_inter(f1, f2, PF_SIZE)
#define pf_diff(f1, f2)        flag_diff(f1, f2, PF_SIZE)

#define player_has(p, flag)       (pf_has(p->race->pflags, (flag)) || pf_has(p->class->pflags, (flag)))


/**
 * The range of possible indexes into tables based upon stats.
 * Currently things range from 3 to 18/220 = 40.
 */
#define STAT_RANGE 38


/**
 * ------------------------------------------------------------------------
 * Player constants
 * ------------------------------------------------------------------------ */


#define PY_MAX_EXP		99999999L	/* Maximum exp */
#define PY_MAX_LEVEL	50			/* Maximum level */

/**
 * Flags for player.spell_flags[]
 */
#define PY_SPELL_LEARNED    0x01 	/* Spell has been learned */
#define PY_SPELL_WORKED     0x02 	/* Spell has been successfully tried */
#define PY_SPELL_FORGOTTEN  0x04 	/* Spell has been forgotten */

#define BTH_PLUS_ADJ    	3 		/* Adjust BTH per plus-to-hit */

/**
 * Player magic realms
 */
enum
{
	#define REALM(a, b, c, d, e, f) REALM_##a,
	#include "list-magic-realms.h"
	#undef REALM
	REALM_MAX
};

/**
 * player noscore flags
 */
#define NOSCORE_WIZARD		0x0002
#define NOSCORE_DEBUG		0x0008
#define NOSCORE_JUMPING     0x0010

/**
 * Skill indexes
 */
enum
{
	SKILL_DISARM,			/* Skill: Disarming */
	SKILL_DEVICE,			/* Skill: Magic Devices */
	SKILL_SAVE,				/* Skill: Saving throw */
	SKILL_STEALTH,			/* Skill: Stealth factor */
	SKILL_SEARCH,			/* Skill: Searching ability */
	SKILL_SEARCH_FREQUENCY,	/* Skill: Searching frequency */
	SKILL_TO_HIT_MELEE,		/* Skill: To hit (normal) */
	SKILL_TO_HIT_BOW,		/* Skill: To hit (shooting) */
	SKILL_TO_HIT_THROW,		/* Skill: To hit (throwing) */
	SKILL_DIGGING,			/* Skill: Digging */

	SKILL_MAX
};

/* Terrain that the player has a chance of digging through */
enum
{
	DIGGING_RUBBLE = 0,
	DIGGING_MAGMA,
	DIGGING_QUARTZ,
	DIGGING_GRANITE,
	DIGGING_DOORS,
	
	DIGGING_MAX
};

/**
 * ------------------------------------------------------------------------
 * Player structures
 * ------------------------------------------------------------------------ */

/**
 * Structure for the "quests"
 */
struct quest
{
	struct quest *next;
	byte index;
	char *name;
	byte level;					/* Dungeon level */
	struct monster_race *race;	/* Monster race */
	int cur_num;				/* Number killed (unused) */
	int max_num;				/* Number required (unused) */
};

/**
 * Player body info
 */
struct equip_slot {
	struct equip_slot *next;

	u16b type;
	char *name;
	struct object *obj;
};

struct player_body {
	struct player_body *next;
	char *name;
	u16b count;
	struct equip_slot *slots;
};

extern struct player_body *bodies;

/**
 * Player gender info
 */
struct player_sex {
	struct player_sex *next;
	const char *name;
	
	unsigned int sidx;    
};

extern struct player_sex *sexes;

/**
 * Player racial info
 */
struct player_race {
	struct player_race *next;
	const char *name;
	
	unsigned int ridx;

	int r_adj[STAT_MAX];	/* Racial stat bonuses */
	
	int r_skills[SKILL_MAX];	/* racial skills */
	
	int r_mhp;			/* Race hit-dice modifier */
	int r_exp;			/* Race experience factor */
	
	int b_age;			/* base age */
	int m_age;			/* mod age */
	
	int base_hgt;		/* base height */
	int mod_hgt;		/* mod height */
	int base_wgt;		/* base weight */
	int mod_wgt;		/* mod weight */
	
	int infra;			/* Infra-vision	range */
	
	int body;			/* Race body */
	struct history_chart *history;
	
	bitflag flags[OF_SIZE];   /* Racial (object) flags */
	bitflag pflags[PF_SIZE];  /* Racial (player) flags */

	struct element_info el_info[ELEM_MAX]; /* Racial resists */
};

extern struct player_race *races;

/**
 * Items the player starts with.  Used in player_class and specified in
 * class.txt.
 */
struct start_item {
	struct object_kind *kind;
	int min;	/* Minimum starting amount */
	int max;	/* Maximum starting amount */

	struct start_item *next;
};


/**
 * Structure for magic realms
 */
struct magic_realm {
	int index;
	int stat;
	const char *verb;
	const char *spell_noun;
	const char *book_noun;
	const char *adjective;
};

extern struct magic_realm realms[REALM_MAX];

/**
 * A structure to hold class-dependent information on spells.
 */
struct class_spell {
	char *name;
	char *text;

	struct effect *effect;	/**< The spell's effect */

	int sidx;		/**< The index of this spell for this class */
	int bidx;		/**< The index into the player's books array */
	int slevel;	/**< Required level (to learn) */
	int smana;		/**< Required mana (to cast) */
	int sfail;		/**< Base chance of failure */
	int sexp;		/**< Encoded experience bonus */
};


/**
 * A structure to hold class-dependent information on spell books.
 */
struct class_book {
	int tval;			/**< Item type of the book */
	int sval;			/**< Item sub-type for book (book number) */
	int realm;			/**< The magic realm of this book */
	int num_spells;	/**< Number of spells in this book */
	struct class_spell *spells;	/**< Spells in the book*/
};


/**
 * Information about class magic knowledge
 */
struct class_magic {
	int spell_first;		/**< Level of first spell */
	int spell_weight;		/**< Max armour weight to avoid mana penalties */
	struct magic_realm *spell_realm;  		/**< Primary spellcasting realm */
	int num_books;			/**< Number of spellbooks */
	struct class_book *books;		/**< Details of spellbooks */
	int total_spells;		/**< Number of spells for this class */
};


/**
 * Player class info
 */
struct player_class {
	struct player_class *next;
	const char *name;
	unsigned int cidx;
	
	const char *title[10];    /* Titles */
	
	int c_adj[STAT_MAX]; /* Class stat modifier */
	
	int c_skills[SKILL_MAX];	/* class skills */
	int x_skills[SKILL_MAX];	/* extra skills */
	
	int c_mhp;        /* Class hit-dice adjustment */
	int c_exp;        /* Class experience factor */
	
	bitflag pflags[PF_SIZE]; /* Class (player) flags */
	
	int max_attacks;  /* Maximum possible attacks */
	int min_weight;   /* Minimum weapon weight for calculations */
	int att_multiply; /* Multiplier for attack calculations */
	
	int sense_base;   /* Base pseudo-id value */
	int sense_div;    /* Pseudo-id divisor */
	
	struct start_item *start_items; /* Starting inventory */
	
	struct class_magic magic; /* Magic spells */
};

extern struct player_class *classes;

/**
 * Histories are a graph of charts; each chart contains a set of individual
 * entries for that chart, and each entry contains a text description and a
 * successor chart to move history generation to.
 * For example:
 * 	chart 1 {
 * 		entry {
 * 			desc "You are the illegitimate and unacknowledged child";
 * 			next 2;
 * 		};
 * 		entry {
 * 			desc "You are the illegitimate but acknowledged child";
 * 			next 2;
 * 		};
 * 		entry {
 * 			desc "You are one of several children";
 * 			next 3;
 * 		};
 * 	};
 *
 * History generation works by walking the graph from the starting chart for
 * each race, picking a random entry (with weighted probability) each time.
 */
struct history_entry {
	struct history_entry *next;
	struct history_chart *succ;
	int isucc;
	int roll;
	char *text;
};

struct history_chart {
	struct history_chart *next;
	struct history_entry *entries;
	unsigned int idx;
};

/**
 * Some more player information
 * This information is retained across player lives
 *
 * XXX - options.c should store most of this, and struct player the rest
 */
typedef struct {
	char full_name[32];		/* Full name */
	
	bool opt[OPT_MAX];		/* Options */
	
	byte hitpoint_warn;		/* Hitpoint warning (0 to 9) */
	u16b lazymove_delay;	/* Delay in cs before moving to allow another key */
	byte delay_factor;		/* Delay factor (0 to 9) */
	
	byte name_suffix;		/* numeric suffix for player name */
} player_other;


/**
 * All the variable state that changes when you put on/take off equipment.
 * Player flags are not currently variable, but useful here so monsters can
 * learn them.
 */
struct player_state {
	s16b speed;		/* Current speed */

	s16b num_blows;		/* Number of blows x100 */
	s16b num_shots;		/* Number of shots */

	byte ammo_mult;		/* Ammo multiplier */
	byte ammo_tval;		/* Ammo variety */

	s16b stat_add[STAT_MAX];	/* Equipment stat bonuses */
	s16b stat_ind[STAT_MAX];	/* Indexes into stat tables */
	s16b stat_use[STAT_MAX];	/* Current modified stats */
	s16b stat_top[STAT_MAX];	/* Maximal modified stats */

	s16b ac;			/* Base ac */
	s16b to_a;			/* Bonus to ac */
	s16b to_h;			/* Bonus to hit */
	s16b to_d;			/* Bonus to dam */

	s16b see_infra;		/* Infravision range */

	s16b cur_light;		/* Radius of light (if any) */

	s16b skills[SKILL_MAX];	/* Skills */

	int noise;			/* Derived from stealth */

	bool heavy_wield;	/* Heavy weapon */
	bool heavy_shoot;	/* Heavy shooter */
	bool icky_wield;	/* Icky weapon shooter */

	bool cumber_armor;	/* Mana draining armor */
	bool cumber_glove;	/* Mana draining gloves */

	bitflag flags[OF_SIZE];	/* Status flags from race and items */
	bitflag pflags[PF_SIZE];	/* Player intrinsic flags */
	struct element_info el_info[ELEM_MAX]; /* Resists from race and items */
};

/**
 * Temporary, derived, player-related variables used during play but not saved
 *
 * Some of these probably should go to the UI
 */
struct player_upkeep {
	bool playing;			/* True if player is playing */
	bool autosave;			/* True if autosave is pending */
	bool generate_level;	/* True if level needs regenerating */
	bool only_partial;		/* True if only partial updates are needed */
	bool dropping;			/* True if auto-drop is in progress */

	int energy_use;			/* Energy use this turn */
	int new_spells;			/* Number of spells available */

	struct monster *health_who;			/* Health bar trackee */
	struct monster_race *monster_race;	/* Monster race trackee */
	struct object *object;				/* Object trackee */
	struct object_kind *object_kind;	/* Object kind trackee */

	u32b notice;		/* Bit flags for pending actions such as 
						 * reordering inventory, ignoring, etc. */
	u32b update;		/* Bit flags for recalculations needed 
						 * such as HP, or visible area */
	u32b redraw;	    /* Bit flags for things that /have/ changed,
						 * and just need to be redrawn by the UI,
						 * such as HP, Speed, etc.*/

	int command_wrk;		/* Used by the UI to decide whether
							 * to start off showing equipment or
							 * inventory listings when offering
							 * a choice.  See obj-ui.c */

	bool create_up_stair;		/* Create up stair on next level */
	bool create_down_stair;		/* Create down stair on next level */

	int resting;				/* Resting counter */

	int running;				/* Running counter */
	bool running_withpathfind;	/* Are we using the pathfinder ? */
	bool running_firststep;		/* Is this our first step running? */

	struct object **quiver;		/* Quiver objects */
	struct object **inven;		/* Inventory objects */
	int total_weight;			/* Total weight being carried */
	int inven_cnt;				/* Number of items in inventory */
	int equip_cnt;				/* Number of items in equipment */
	int quiver_cnt;				/* Number of items in the quiver */
};


/**
 * Most of the "player" information goes here.
 *
 * This stucture gives us a large collection of player variables.
 *
 * This entire structure is wiped when a new character is born.
 *
 * This structure is more or less laid out so that the information
 * which must be saved in the savefile precedes all the information
 * which can be recomputed as needed.
 */
struct player {
	s16b py;			/* Player location */
	s16b px;			/* Player location */

        byte psex;			/* Sex index */
        
        const struct player_sex *sex;
	const struct player_race *race;
	const struct player_class *class;

	byte hitdie;		/* Hit dice (sides) */
	byte expfact;		/* Experience factor */

	s16b age;			/* Characters age */
	s16b ht;			/* Height */
	s16b wt;			/* Weight */

	s32b au;			/* Current Gold */

	s16b max_depth;		/* Max depth */
	s16b depth;			/* Cur depth */

	s16b max_lev;		/* Max level */
	s16b lev;			/* Cur level */

	s32b max_exp;		/* Max experience */
	s32b exp;			/* Cur experience */
	u16b exp_frac;		/* Cur exp frac (times 2^16) */

	s16b mhp;			/* Max hit pts */
	s16b chp;			/* Cur hit pts */
	u16b chp_frac;		/* Cur hit frac (times 2^16) */

	s16b msp;			/* Max mana pts */
	s16b csp;			/* Cur mana pts */
	u16b csp_frac;		/* Cur mana frac (times 2^16) */

	s16b stat_max[STAT_MAX];	/* Current "maximal" stat values */
	s16b stat_cur[STAT_MAX];	/* Current "natural" stat values */

	s16b *timed;		/* Timed effects */

	s16b word_recall;	/* Word of recall counter */
	s16b deep_descent;	/* Deep Descent counter */

	s16b energy;		/* Current energy */
	u32b total_energy;	/* Total energy used (including resting) */
	u32b resting_turn;	/* Number of player turns spent resting */

	s16b food;			/* Current nutrition */

	byte confusing;		/* Glowing hands */
	byte searching;		/* Currently searching */
	byte unignoring;	/* Unignoring */

	byte *spell_flags; /* Spell flags */
	byte *spell_order;	/* Spell order */

	s16b player_hp[PY_MAX_LEVEL];	/* HP Array */

	char died_from[80];		/* Cause of death */
	char *history;			/* Player history */
	struct quest *quests;	/* Quest history */
	u16b total_winner;		/* Total winner */

	u16b noscore;			/* Cheating flags */

	bool is_dead;			/* Player is dead */

	bool wizard;			/* Player is in wizard mode */

	/* Generation fields (for quick start) */
	s32b au_birth;          /* Birth gold when option birth_money is false */
	s16b stat_birth[STAT_MAX]; /* Birth "natural" stat values */
	s16b ht_birth;          /* Birth Height */
	s16b wt_birth;          /* Birth Weight */

	/* Variable and calculatable player state */
	struct player_state state;
	struct player_state known_state;

	/* Tracking of various temporary player-related values */
	struct player_upkeep *upkeep;

	/* Real gear */
	struct object *gear;
	/* Known gear */
	struct object *gear_k;

	struct player_body body;
};


/**
 * ------------------------------------------------------------------------
 * Externs
 * ------------------------------------------------------------------------ */

extern const s32b player_exp[PY_MAX_LEVEL];
extern player_other *op_ptr;
extern struct player *player;


/* player-class.c */
extern struct player_class *player_id2class(guid id);

/* player.c */
int stat_name_to_idx(const char *name);
const char *stat_idx_to_name(int type);
extern bool player_stat_inc(struct player *p, int stat);
extern bool player_stat_dec(struct player *p, int stat, bool permanent);
extern void player_exp_gain(struct player *p, s32b amount);
extern void player_exp_lose(struct player *p, s32b amount, bool permanent);
extern void player_flags(struct player *p, bitflag f[OF_SIZE]);

extern byte player_hp_attr(struct player *p);
extern byte player_sp_attr(struct player *p);

extern bool player_restore_mana(struct player *p, int amt);

extern const char *player_safe_name(struct player *p, bool strip_suffix);

/* player-race.c */
extern struct player_race *player_id2race(guid id);

#endif /* !PLAYER_H */
